jwst/pixel_replace/pixel_replace.py

```diff
@@ -0,0 +1,382 @@
+import logging
+import numpy as np
+from .. import datamodels
+from scipy.optimize import minimize
+import warnings
+from ..assign_wcs.nirspec import nrs_wcs_set_input
+
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+
+class PixelReplacement:
+    """Main class for performing pixel replacement.
+
+    This class controls loading the input data model, selecting the
+    method for pixel replacement, and executing each step. This class
+    should provide modularization to allow for multiple options and possible
+    future reference files.
+
+    """
+
+    # Shortcuts for DQ Flags
+    DO_NOT_USE = datamodels.dqflags.pixel['DO_NOT_USE']
+    FLUX_ESTIMATED = datamodels.dqflags.pixel['FLUX_ESTIMATED']
+    NON_SCIENCE = datamodels.dqflags.pixel['NON_SCIENCE']
+
+    # Shortcuts for dispersion direction for ease of reading
+    HORIZONTAL = 1
+    VERTICAL = 2
+
+    default_suffix = 'pixrep'
+
+    def __init__(self, input_model, **pars):
+        """
+        Initialize the class with input data model.
+
+        Parameters
+        ----------
+        input_model : DataModel, str
+            list of data models as ModelContainer or ASN file,
+            one data model for each input image
+
+        pars : dict, optional
+            Optional parameters to modify how pixel replacement
+            will execute.
+        """
+        self.input = input_model
+        self.pars = dict()
+        self.pars.update(pars)
+        self.output = self.input.copy()
+        # Store algorithm options here.
+        self.algorithm_dict = {
+            'fit_profile': self.fit_profile,
+        }
+
+        # Choose algorithm from dict using input par.
+        try:
+            self.algorithm = self.algorithm_dict[self.pars['algorithm']]
+
+        except KeyError:
+            log.critical(f"Algorithm name {self.pars['algorithm']} provided does "
+                         f"not match an implemented algorithm!")
+            raise Exception
+
+    def replace(self):
+        """
+        Process the input DataModel, unpack any model that holds
+        more than one 2D spectrum, then apply selected algorithm
+        to each 2D spectrum in input.
+        """
+        # ImageModel inputs (MIR_LRS-FIXEDSLIT)
+        if isinstance(self.input, datamodels.ImageModel):
+            self.output = self.algorithm(self.input)
+            n_replaced = np.count_nonzero(self.output.dq & self.FLUX_ESTIMATED)
+            log.info(f"Input model had {n_replaced} pixels replaced.")
+        elif isinstance(self.input, datamodels.IFUImageModel):
+                # Attempt to run pixel replacement on each throw of the IFU slicer
+                # individually.
+                xx, yy = np.indices(self.input.data.shape)
+                if self.input.meta.exposure.type == 'MIR_MRS':
```

Comment:
It may be - I'm not entirely sure. My thought process was that this handler function would "massage" the datamodels to provide each algorithm a consistent input format - for most datamodels this is straightforward, but for IFU data this required a lot more massaging. It's possible that this all gets excised if the IFU instruments decide this should never be applied to the data, but it felt like a waste to delete all of my hard work. 🙂